csrc/optimization/opt_pass.cpp

```diff
@@ -0,0 +1,67 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <optimization/opt_pass.h>
+
+#include <mutex>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+class OptimizationRegistry {
+ public:
+  struct PassEntry {
+    int priority_;
+    FusionPass pass_;
+    std::string name_;
+    PassEntry(int priority, FusionPass pass, std::string name) : priority_(priority), pass_(std::move(pass)), name_(std::move(name_)) {}
+  };
+
+  void registerPass(const OptimizationPassCategory& cat, FusionPass func, std::string name_, int priority) {
+    std::lock_guard<std::mutex> guard(mutex_);
```

Comment:
Is this going to be multithreaded?

Reply:
I'm using a static registry, so this is just trying to make it thread-safe.

We can try to improve thread efficiency. i.e. currently read/write shares a global mutex and that's unnecessarily inefficient. If concurrent calls are prevalent.

Reply:
Let me rephrase my question. Is this path multithreaded? I generally assumed we are single-threaded, except for the parallel nvrtc compilation. Is it a valid assumption?

Reply:
> Is this path multithreaded?

Yes. It could be.

> I generally assumed we are single-threaded, except for the parallel nvrtc compilation.

That's a tricky question. Generally speaking, we do try to be thread-safe (look at our [thread_local fusion guard](https://github.com/NVIDIA/Fuser/blob/454d020590279202f89a7cc9af8d49f1d4b58a3d/csrc/fusion.cpp#L28)). though in reality, we don't do a good job there. Historically, that's largely due to most pytorch workflow are somewhat single-threaded.

Reply:
The fusion guard is just thread local, so there's no sharing of data. Here, since you're using a mutex, I'm assuming it is to prevent data races of some shared data. When does this sharing happen?

Reply:
I still don't understand when the data sharing happens.

Reply:
Oh sorry missed the previous message. We are using a singleton here. Look at `getInstance` defined below.

Reply:
We don't have a use case in our code base.

My motivation is that, user/frontend might have separate threads using nvfuser.
I'm assuming that we want a single point to define/register passes. Hence making the register `static` instead of `thread_local`, which calls for a mutex to guard it.

Reply:
Oh, so you are thinking about this could be called from outside nvfuser.

Is this going to be no longer an issue after [this](https://github.com/NVIDIA/Fuser/pull/355#discussion_r1198500722)?

Reply:
Even after the refactor on registration, it might still be an issue. We agreed to refactor the registration to not be done via static initialization so we can have a well defined order of passes with identical priority.

Maybe we are not totally on the same page how/where the grouped passes is going to be registered :fearful: 
I'm still viewing that there will be a global (static storage?!) registry somewhere, and optimization passes can be changed at runtime. Now I'm getting the sense that this part isn't what you want to support and you are leaning towards a total compile time pass composition?!

Reply:
>how/where the grouped passes is going to be registered

I actually still don't know what it should look like as I don't have a clear idea of how it should be exposed to the user. In fact, we don't have a well defined interface to customize the whole pass of segmentation, lowering and compilation, at all.

I wonder if we shouldn't worry too much about exposing the customization of this optimization and rather just have a compile-time composition. This isn't saying we should not provide any customization interface, but I'd prefer starting with something simpler as we should spend more time to think about what the interface should look like.

Reply:
Sorry for not keeping my comment up-to-date.

> I wonder if we shouldn't worry too much about exposing the customization of this optimization and rather just have a compile-time composition.

I agree with you. I can't justify having complex logic in the code without a real-life use-case. I actually already switched the implementation back to this. If you look at the `pre_segmenter.cpp`, it's just a plain stitching of existing passes now.